python/cog/server/redis_queue.py

```diff
@@ -189,80 +189,49 @@ def start(self) -> None:
-                        self.redis.xadd(
-                            self.predict_time_queue,
-                            fields={"duration": run_time},
-                            maxlen=self.stats_queue_length,
-                        )
-                        sys.stderr.write(f"Run time for {message_id}: {run_time:.2f}\n")
-                    except Exception as e:
-                        response["status"] = Status.FAILED
-                        response["error"] = str(e)
-                        response["completed_at"] = format_datetime(
-                            datetime.datetime.now()
-                        )
+                    for response in self.run_prediction(message, should_cancel):
                         send_response(response)
-                        self.redis.xack(self.input_queue, self.input_queue, message_id)
-                        self.redis.xdel(self.input_queue, message_id)
-                    finally:
-                        for cleanup_function in cleanup_functions:
-                            try:
-                                cleanup_function()
-                            except Exception as e:
-                                sys.stderr.write(f"Cleanup function caught error: {e}")
+
+                    self.redis.xack(self.input_queue, self.input_queue, message_id)
+                    self.redis.xdel(self.input_queue, message_id)
+
             except Exception as e:
                 tb = traceback.format_exc()
                 sys.stderr.write(f"Failed to handle message: {tb}\n")
 
         sys.stderr.write("Shutting down worker: bye bye!\n")
         self.worker.shutdown()
 
-    def handle_message(
-        self,
-        send_response: Callable,
-        response: Dict[str, Any],
-        cleanup_functions: List[Callable],
-        should_cancel: Callable,
-    ) -> None:
-        span = trace.get_current_span()
+    def run_prediction(
+        self, message: Dict[str, Any], should_cancel: Callable
+    ) -> Iterable[Dict[str, Any]]:
+        # use the request message as the basis of our response so
+        # that we echo back any additional fields sent to us
+        response = message
+        response["status"] = Status.PROCESSING
+        response["output"] = None
+        response["logs"] = ""
+
         started_at = datetime.datetime.now()
 
         try:
             input_obj = self.InputType(**response["input"])
-        except ValidationError as e:
-            tb = traceback.format_exc()
-            sys.stderr.write(tb)
+        except Exception as e:
             response["status"] = Status.FAILED
             response["error"] = str(e)
-            send_response(response)
-            span.record_exception(e)
-            span.set_status(TraceStatus(status_code=StatusCode.ERROR))
-            return
+            yield response
 
-        cleanup_functions.append(input_obj.cleanup)
+            try:
+                input_obj.cleanup()
```

Comment:
I'm not sure I know what the implications are of calling this before we pass `input_obj` into `.predict(...)`. Is that guaranteed to be safe?

Reply:
This is only called if input validation fails, and we're about to bail out.

Reply:
It definitely shows it could be clearer! Tbh, the main reason I didn't do go further on inputs is because we talked about moving that inside the predictor anyway, at which point this code is going away.